File: voucherc/src/features.rs
use crate::types::Feature;
use static_init::dynamic;
use std::collections::HashSet;

#[dynamic]
static DICT: HashSet<String> = include_str!("../data/dictionary.en.txt")
    .lines()
    .map(ToOwned::to_owned)
    .collect();

#[allow(dead_code)]
pub fn from_word(w: &str) -> Feature {
    let b2f = |b| if b { 1.0 } else { 0.0 };

    vec![
        w.len() as f64,
        b2f(is_lowercase(&w)),
        b2f(is_uppercase(&w)),
        b2f(has_letters(&w)),
        b2f(has_letters_only(&w)),
        b2f(has_digits(&w)),
        b2f(has_digits_only(&w)),
        b2f(are_more_than_half_digits(&w)),
        b2f(is_alphanumeric_or_dash_or_underscore(&w)),
        b2f(has_letters_which_end_with_two_digits(&w)),
        b2f(ends_with_digits(&w)),
        b2f(is_in_english_dictionary(&w)),
    ]
}

fn is_lowercase(w: &str) -> bool {
    w.matches(char::is_uppercase).next().is_none()
        && w.matches(char::is_lowercase).next().is_some()
}

fn is_uppercase(w: &str) -> bool {
    w.matches(char::is_uppercase).next().is_some()
        && w.matches(char::is_lowercase).next().is_none()
}

fn has_letters(w: &str) -> bool {
    w.chars().any(char::is_alphabetic)
}

fn has_letters_only(w: &str) -> bool {
    !w.is_empty() && w.chars().all(char::is_alphabetic)
}

fn has_digits(w: &str) -> bool {
    w.chars().any(char::is_numeric)
}

fn has_digits_only(w: &str) -> bool {
    !w.is_empty() && w.chars().all(char::is_numeric)
}

fn are_more_than_half_digits(w: &str) -> bool {
    w.len() / 2 < w.chars().filter(|c| char::is_numeric(*c)).count()
}

fn is_alphanumeric_or_dash_or_underscore(w: &str) -> bool {
    !w.is_empty()
        && w.chars()
            .all(|c| char::is_alphanumeric(c) || c == '_' || c == '-')
}

fn ends_with_digits(w: &str) -> bool {
    w.chars().last().filter(|c| char::is_numeric(*c)).is_some()
}

fn has_letters_which_end_with_two_digits(w: &str) -> bool {
    let mut chars = w.chars().rev().peekable();
    chars.next().filter(|c| char::is_numeric(*c)).is_some()
        && chars.next().filter(|c| char::is_numeric(*c)).is_some()
        && chars.peek().is_some()
        && chars.all(char::is_alphabetic)
}

fn is_in_english_dictionary(w: &str) -> bool {
    DICT.contains(&w.to_lowercase())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_lowercase() {
        assert!(is_lowercase("this123-_"));
        assert!(!is_lowercase("thIs123-_"));
        assert!(!is_lowercase("123-_"));
        assert!(!is_lowercase("AHOJ123-_"));
        assert!(!is_lowercase(""));
    }

    #[test]
    fn test_is_uppercase() {
        assert!(is_uppercase("THIS123-_"));
        assert!(!is_uppercase("thIs123-_"));
        assert!(!is_uppercase("123-_"));
        assert!(!is_uppercase("ahoj123-_"));
        assert!(!is_uppercase(""));
    }

    #[test]
    fn test_has_letters_only() {
        assert!(has_letters_only("onlyleTTErs"));
        assert!(!has_letters_only("123TErs"));
        assert!(!has_letters_only(""));
        assert!(!has_letters_only("-_"));
        assert!(!has_letters_only("123"));
    }

    #[test]
    fn test_has_letters() {
        assert!(has_letters("onlyleTTErs"));
        assert!(has_letters("123TErs"));
        assert!(!has_letters(""));
        assert!(!has_letters("-_"));
        assert!(!has_letters("123"));
    }

    #[test]
    fn test_has_digits_only() {
        assert!(has_digits_only("123"));
        assert!(!has_digits_only("onlyleTTErs"));
        assert!(!has_digits_only("123TErs"));
        assert!(!has_digits_only(""));
        assert!(!has_digits_only("-_"));
    }

    #[test]
    fn test_has_digits() {
        assert!(has_digits("123"));
        assert!(has_digits("123TErs"));
        assert!(!has_digits("onlyleTTErs"));
        assert!(!has_digits(""));
        assert!(!has_digits("-_"));
    }

    #[test]
    fn test_are_more_than_half_digits() {
        assert!(are_more_than_half_digits("123"));
        assert!(are_more_than_half_digits("1234abc"));
        assert!(!are_more_than_half_digits("at_123"));
        assert!(!are_more_than_half_digits("onlyleTTErs"));
        assert!(!are_more_than_half_digits(""));
        assert!(!are_more_than_half_digits("1a"));
    }

    #[test]
    fn test_is_alphanumeric_or_dash_or_underscore() {
        assert!(is_alphanumeric_or_dash_or_underscore("123"));
        assert!(is_alphanumeric_or_dash_or_underscore("1234abc"));
        assert!(is_alphanumeric_or_dash_or_underscore("at_123"));
        assert!(is_alphanumeric_or_dash_or_underscore("-123-_onlyleTTErs"));
        assert!(!is_alphanumeric_or_dash_or_underscore(""));
        assert!(!is_alphanumeric_or_dash_or_underscore("1a$"));
    }

    #[test]
    fn test_ends_with_digits() {
        assert!(ends_with_digits("123"));
        assert!(ends_with_digits("at_123"));
        assert!(!ends_with_digits("1234abc"));
        assert!(!ends_with_digits("-123-_onlyleTTErs"));
        assert!(!ends_with_digits(""));
        assert!(!ends_with_digits("1a$"));
    }

    #[test]
    fn test_has_letters_which_end_with_two_digits() {
        assert!(has_letters_which_end_with_two_digits("OFF20"));
        assert!(!has_letters_which_end_with_two_digits("123hehe12"));
        assert!(!has_letters_which_end_with_two_digits("23"));
        assert!(!has_letters_which_end_with_two_digits("123"));
        assert!(!has_letters_which_end_with_two_digits("123hehe1"));
    }

    #[test]
    fn test_is_in_english_dictionary() {
        assert!(is_in_english_dictionary("dog"));
        assert!(is_in_english_dictionary("DOG"));
        assert!(!is_in_english_dictionary("d4wg"));
        assert!(!is_in_english_dictionary("nepravdepodobne"));
    }
}

File: voucherc/features.py
import re


def from_str(w):
    """
    Given a string it converts it to a list of floats.
    """
    assert isinstance(w, str), "Input must be string"
    assert len(w) != 0, "Input cannot be empty"

    features = [
        float(len(w)),
        float(w.islower()),
        float(w.isupper()),
        float(has_letters(w)),
        float(has_letters_only(w)),
        float(has_digits(w)),
        float(has_digits_only(w)),
        float(is_alphanumeric_or_dash_or_underscore(w)),
        float(ends_with_digit(w)),
        float(has_letters_which_end_with_two_digits(w)),
        float(is_in_english_dictionary(w)),
    ]

    for i in range(0, len(w)):
        features.append(map_char_to_feature(w[i]))

    return features


def has_letters(w):
    return bool(re.search("[a-zA-Z]", w))


def has_letters_only(w):
    return bool(re.search("^[a-zA-Z]+$", w))


def has_digits(w):
    return bool(re.search("\d", w))


def has_digits_only(w):
    return bool(re.search("^\d+$", w))


def is_alphanumeric_or_dash_or_underscore(w):
    return bool(re.search("^[a-zA-Z0-9-_]+$", w))


def ends_with_digit(w):
    return ord(w[-1]) in range(ord("0"), ord("9") + 1)


def has_letters_which_end_with_two_digits(w):
    return bool(re.search("^[a-zA-Z]+\d\d$", w))


def load_dict_en():
    """
    Loads dictionary from file. This should happen once on program boot.
    """
    print("Loading dict")
    dict_en = open("data/dictionary.en.txt", "r")

    words = set()

    for word in dict_en.readlines():
        words.add(word.strip().lower())

    return words


dict_en = load_dict_en()


def is_in_english_dictionary(w):
    return w.lower() in dict_en


def map_char_to_feature(c):
    c = ord(c)
    if c in range(ord("a"), ord("z") + 1):
        return 1.0
    if c in range(ord("A"), ord("Z") + 1):
        return 2.0
    if c in range(ord("0"), ord("9") + 1):
        return 3.0
    if c == ord("-"):
        return 4.0

    return 5.0


"""
Uncomment last line to run tests
"""


def test_module():
    assert has_letters("$_THERE")
    assert has_letters("there_123")
    assert not has_letters("123(*&^")

    assert has_letters_only("THERE")
    assert has_letters_only("there")
    assert not has_letters_only("there_123")

    assert has_digits("yes_123")
    assert not has_digits("no:(")

    assert has_digits_only("80082")
    assert not has_digits_only("no_123")

    assert is_alphanumeric_or_dash_or_underscore("123")
    assert is_alphanumeric_or_dash_or_underscore("1234abc")
    assert is_alphanumeric_or_dash_or_underscore("at_123")
    assert is_alphanumeric_or_dash_or_underscore("-123-_onlyleTTErs")
    assert not is_alphanumeric_or_dash_or_underscore("")
    assert not is_alphanumeric_or_dash_or_underscore("1a$")

    assert ends_with_digit("123")
    assert ends_with_digit("at_129")
    assert ends_with_digit("at_120")
    assert not ends_with_digit("1234abc")

    assert has_letters_which_end_with_two_digits("OFF20")
    assert not has_letters_which_end_with_two_digits("123hehe12")
    assert not has_letters_which_end_with_two_digits("23")
    assert not has_letters_which_end_with_two_digits("123")
    assert not has_letters_which_end_with_two_digits("123hehe1")

    assert is_in_english_dictionary("dog")
    assert is_in_english_dictionary("DOG")
    assert not is_in_english_dictionary("d4wg")
    assert not is_in_english_dictionary("nepravdepodobne")

    assert map_char_to_feature("a") == 1.0
    assert map_char_to_feature("z") == 1.0
    assert map_char_to_feature("A") == 2.0
    assert map_char_to_feature("Z") == 2.0
    assert map_char_to_feature("0") == 3.0
    assert map_char_to_feature("1") == 3.0
    assert map_char_to_feature("9") == 3.0
    assert map_char_to_feature("-") == 4.0
    assert map_char_to_feature("$") == 5.0
    assert map_char_to_feature("%") == 5.0

    assert from_str("OK20") == [
        4.0,
        0.0,
        1.0,
        1.0,
        0.0,
        1.0,
        0.0,
        1.0,
        1.0,
        1.0,
        0.0,
        2.0,
        2.0,
        3.0,
        3.0,
    ]


# uncomment temporarily when working with module
# test_module()
